File: cplus/leetcode/202501/algo1414.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 1e9;
vector<int> fib;
int init = [] {
    fib.push_back(1), fib.push_back(1);
    while (fib[fib.size()-1] < N)
        fib.push_back(fib[fib.size()-1] + fib[fib.size()-2]);
    return 0;
}();

// 贪心专题 - 其他数学贪心
class Solution {
public:
    int findMinFibonacciNumbers(int k) {
        int ans = 0, inx = fib.size() - 1;
        while (k) {
            while (fib[inx] > k) inx -= 1;
            ans ++, k -= fib[inx];
        }
        return ans;
    }
};

File: python/algo/202501/algo1414.py
from bisect import bisect_right


N = 10 ** 9
fib = [1,1]
while fib[-1] < N:
    fib.append(fib[-1]+fib[-2])

class Solution:
    def findMinFibonacciNumbers(self, k: int) -> int:
        ans = 0
        while k:
            inx = bisect_right(fib, k)
            ans += 1
            k -= fib[inx-1]

        return ans    
        